subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiJdkCompatibilityTest.groovy

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2011 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.integtests.tooling
+
+import org.gradle.api.JavaVersion
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.AvailableJavaHomes
+import org.gradle.integtests.fixtures.executer.GradleContextualExecuter
+import org.gradle.internal.jvm.Jvm
+import org.gradle.test.fixtures.file.TestFile
+import org.gradle.util.Requires
+import org.gradle.util.TestPrecondition
+import org.junit.Assume
+import spock.lang.Unroll
+
+class ToolingApiJdkCompatibilityTest extends AbstractIntegrationSpec {
+
+    TestFile projectDir
+
+    def setup() {
+        projectDir = temporaryFolder.testDirectory
+        temporaryFolder.testDirectory.file("build.gradle") << ""
+        temporaryFolder.testDirectory.file("settings.gradle") << "rootProject.name = 'target-project'"
+    }
+
+    @Unroll
+    @Requires(adhoc = { TestPrecondition.JDK11_OR_LATER.fulfilled && !GradleContextualExecuter.embedded })
+    def "Java #compilerJdkVersion.majorVersion client can launch task on Java #clientJdkVersion.majorVersion with Gradle #gradleVersion on Java #gradleDaemonJdkVersion.majorVersion"(JavaVersion compilerJdkVersion, JavaVersion clientJdkVersion, JavaVersion gradleDaemonJdkVersion, String gradleVersion) {
```

Comment:
This is close to what I was thinking we could do, but I think we should avoid running the compiler ourselves.  I also think we should run this test regardless if the test is running on JDK8, JDK11 or beyond.

Take a look at this reworking of this test:
https://github.com/gradle/gradle/commit/89e981a0ec32563057891502ec1c7ccbcfd1707b

Instead of compiling the TAPI client by hand, everything is wrapped into a Gradle build that can do a few things for us:
- Resolve the TAPI client dependencies (vs relying on the classpath of the test to have it)
- Find a JDK8 toolchain so we can build 6-8 compatible classes
- Target the right compatibility level

I think this makes it a little easier to also compile a custom action since everything is contained in the 'client-runner' project.

PTAL and see if this is helpful.

Reply:
It is absolutely helpful, thanks Sterling, for making the modeling much nicer.